### PR TITLE
Feature.mgmtint

### PR DESCRIPTION
**Pull request by jasonrahm**
updated test cases to reflect jenkins environment

### CODE REVIEW

**Comment by jasonrahm** [comment]
@wojtek0806 or @caphrim007 this is ready for review. Note that the create method doesn't fully test a working create and the delete method is not tested to protect the management interface connectivity for all other tests. That logic is there, just commented out.

File: f5/bigip/tm/sys/management_ip.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# coding=utf-8
+#
+# Copyright 2016 F5 Networks Inc.
```
Comment (caphrim007) [inline f5/bigip/tm/sys/management_ip.py]: we're now in 2017

File: f5/bigip/tm/sys/management_route.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# coding=utf-8
+#
+# Copyright 2016 F5 Networks Inc.
```
Comment (caphrim007) [inline f5/bigip/tm/sys/management_route.py]: year

File: f5/bigip/tm/sys/test/functional/test_management_ip.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
+# Copyright 2015-2106 F5 Networks Inc.
```
Comment (caphrim007) [inline f5/bigip/tm/sys/test/functional/test_management_ip.py]: same year. also, needs the python shebang at the top

**Comment by caphrim007** [comment]
@jasonrahm can you really not update the management ip through the api?

File: f5/bigip/tm/sys/management_route.py (hunk 2 above)
Comment (jasonrahm) [reply f5/bigip/tm/sys/management_route.py]: heh, I suppose I should read comment fields I copy over from other files more carefully...I can clean up here or I can submit a PR to update all source files for copyright info.

Comment (caphrim007) [reply f5/bigip/tm/sys/management_route.py]: we dont need to update all source files in the source tree for copyright, just the new files that we add.

**Comment by jasonrahm** [comment]
You can update the management IP but it is done through create, not modify or update.

**Comment by caphrim007** [comment]
so you can create and you can delete, and updating is no different from creating?

**Comment by jasonrahm** [comment]
A POST will overwrite existing management IP as long as it is different than the existing one. PUT/PATCH will not do anything to the management IP.

**Comment by caphrim007** [comment]
I'm wondering then if, for consistency and user experience's sake, if we should just alias modify commands to create, and hide that implementation detail from the user

File: f5/bigip/tm/sys/test/functional/test_management_ip.py (hunk 3 above)
Comment (jasonrahm) [reply f5/bigip/tm/sys/test/functional/test_management_ip.py]: when did we start adding the shebang to files? I haven't seen that in the other modules

**Comment by jasonrahm** [comment]
do you mean that we should only support .update() or .modify() , but in those, call the create method instead?

**Comment by caphrim007** [comment]
Ignore my shebang comment. You're correct.

For modify and update, yes, if it behaves where the experience of "updating" is essentially just calling create, then i would think the user would think they are updating, when in dact we're just calling create on their behalf

**Comment by jasonrahm** [comment]
ok i will refactor, thanks

src: mobile

On Feb 15, 2017, at 7:28 PM, Tim Rupp <notifications@github.com<mailto:notifications@github.com>> wrote:


Ignore my shebang comment. You're correct.

For modify and update, yes, if it behaves where the experience of "updating" is essentially just calling create, then i would think the user would think they are updating, when in dact we're just calling create on their behalf

-

**Comment by wojtek0806** [comment]
what happens if you don't have mgmt IP set?, i think it would be better if we leave create(), and just redo modify() or update() or both to use POST. This way this will be cleaner than aliasing and doing all weird voodoo stuff

@jasonrahm

**Comment by jasonrahm** [comment]
yes, create needs to still be there for initial setting of management interface if absent.

**Comment by wojtek0806** [comment]
just override _update() method and make_modify() not allowed

File: f5/bigip/tm/sys/management_ip.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+            {'tm:sys:management-ip:management-ipstate': Management_Ip}
+
+
+class Management_Ip(Resource):
+    """BIG-IP® management-ip resource"""
+    def __init__(self, Management_Ips):
+        super(Management_Ip, self).__init__(Management_Ips)
+        self._meta_data['required_json_kind'] = \
+            'tm:sys:management-ip:management-ipstate'
+        self._meta_data['required_creation_parameters'].update(('name',))
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/management_ip.py]: that should be removed as we already have this as default in Resource super class

File: f5/bigip/tm/sys/management_route.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        self._meta_data['attribute_registry'] =\
+            {'tm:sys:management-route:management-routestate': Management_Route}
+
+
+class Management_Route(Resource):
+    """BIG-IP® system management-route resource"""
+    def __init__(self, Management_Routes):
+        super(Management_Route, self).__init__(Management_Routes)
+        self._meta_data['required_creation_parameters'].update(
+            ('name', 'network', 'gateway'))
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/management_route.py]: again you can leave out the 'name' parameter as this is already there.  Should be something like this instead:

` self._meta_data['required_creation_parameters'].update(('network', 'gateway'))`

File: f5/bigip/tm/sys/management_ip.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+    def load(self, **kwargs):
+        kwargs['transform_name'] = True
+        return self._load(**kwargs)
+
+    def modify(self, **kwargs):
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/management_ip.py]: not entirely true, there is a property called 'description', and 'app-service' which can be manipulated, so this should be allowed

File: f5/bigip/tm/sys/management_ip.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        raise UnsupportedMethod(
+            "%s does not support the modify method" % self.__class__.__name__)
+
+    def update(self, **kwargs):
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/management_ip.py]: not entirely true, there is a property called 'description', and 'app-service' which can be manipulated, so this should be allowed

File: f5/bigip/tm/sys/test/functional/test_management_ip.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+#         mgmt_root.tm.sys.management_ips.management_ip.load(
+#             name=mip[0].name)
+#     assert ') was not found.' in str(err.value.message)
+#
+#     mip2 = mgmt_root.tm.sys.management_ips.management_ip.create(
+#         name='172.16.44.16/24')
+#     assert mip2.name == '172.16.44.16/24'
+
+
+def test_modify_addr(request, mgmt_root):
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/test/functional/test_management_ip.py]: this needs changing due to update/modify being possible

File: f5/bigip/tm/sys/test/functional/test_management_ip.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    name = '10.0.2.15/24'
+    mip1 = mgmt_root.tm.sys.management_ips.management_ip.load(
+        name=name)
+    mip1.name = '10.0.2.16/24'
+    with pytest.raises(UnsupportedMethod) as err:
+        mip1.modify()
+    assert 'Management_Ip does not support the modify method' in str(err)
+
+
+def test_update_addr(request, mgmt_root):
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/test/functional/test_management_ip.py]: this needs changing do to the update being possible

**Review by wojtek0806 — CHANGES_REQUESTED** [review]
See my inline comments

**Review by wojtek0806 — APPROVED** [review]

**Review by wojtek0806 — CHANGES_REQUESTED** [review]
see my inline comments and correct accordingly

File: f5/bigip/tm/sys/management_ip.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    """BIG-IP® management-ip resource"""
+    def __init__(self, Management_Ips):
+        super(Management_Ip, self).__init__(Management_Ips)
+        self._meta_data['required_json_kind'] = \
+            'tm:sys:management-ip:management-ipstate'
+        self._meta_data['required_creation_parameters'].update(('name',))
+
+    def load(self, **kwargs):
+        kwargs['transform_name'] = True
+        return self._load(**kwargs)
```
Comment (wojtek0806) [inline f5/bigip/tm/sys/management_ip.py]: nice!

**Comment by jasonrahm** [comment]
ready for your re-review @wojtek0806

**Review by wojtek0806 — APPROVED** [review]
Looks good now, we can add 'create' test later, for Lack of 2 NAT interfaces inside VirtualBox in Vagrant is not helping. I need to talk to @caphrim007 about creating and using NAT network, this way you can have multiple interfaces with NAT on them

**Pull request merged**